assume/markets/clearing_algorithms/complex_clearing.py

```diff
@@ -136,114 +136,128 @@ def energy_balance_rule(m, t):
     return instance, results
 
 
-def pay_as_clear_complex(
-    market_agent: MarketRole,
-    market_products: list[MarketProduct],
-):
-    """
-    This implements pay-as-clear with more complex bid structures, including acceptance ratios, bid types, and profiled volumes.
+class ComplexClearingRole(MarketRole, MarketMechanism):
+    def __init__(self, marketconfig: MarketConfig):
+        super().__init__(marketconfig)
+        assert "bid_type" in self.marketconfig.additional_fields
 
-    :param market_agent: The market agent
-    :type market_agent: MarketRole
-    :param market_products: The products to be traded
-    :type market_products: list[MarketProduct]
+    def validate_orderbook(self, orderbook: Orderbook, agent_tuple) -> None:
+        super().validate_orderbook(orderbook, agent_tuple)
+        for order in orderbook:
+            order["bid_type"] = "SB" if order["bid_type"] == None else order["bid_type"]
+            assert order["bid_type"] in [
+                "SB",
+                "BB",
+            ], f"bid_type {order['bid_type']} not in ['SB', 'BB']"
 
-    :return extract_results(model=model, eps=eps, orders=orders, market_products=market_products, market_clearing_prices=market_clearing_prices)
-    :rtype: tuple[Orderbook, Orderbook, list[dict]]
-    """
+            if order["bid_type"] == "BB":
+                assert False not in [
+                    abs(volume) <= max_volume for _, volume in order["volume"].items()
```

Comment:
max_volume is not defined, where is it comming from?

Reply:
true, weird that this is not triggered by tests.
`max_volume = self.marketconfig.maximum_bid_volume`